packages/utils/src/cryptography/merkle_proof.cairo

```diff
@@ -0,0 +1,149 @@
+/// that: 1) the tree is complete (but not necessarily perfect), 2) the leaves to be proven are in
+/// the opposite order they are in the tree (i.e., as seen from right to left starting at the
+/// deepest layer and continuing at the next layer).
+///
+/// NOTE: The _empty set_ (i.e. the case where `proof.len() == 1 && leaves.len() == 0`) is
+/// considered a no-op, and therefore a valid multiproof (i.e. it returns `proof.at(0)`). Consider
+/// disallowing this case if you're not validating the leaves elsewhere.
+pub fn process_multi_proof<impl Hasher: CommutativeHasher>(
+    proof: Span<felt252>, proof_flags: Span<bool>, leaves: Span<felt252>
+) -> felt252 {
+    // This function rebuilds the root hash by traversing the tree up from the leaves. The root is
+    // rebuilt by consuming and producing values on a queue. The queue starts with the `leaves`
+    // span, then goes onto the `hashes` span. At the end of the process, the last hash in the
+    // `hashes` span should contain the root of the Merkle tree.
+    let leaves_len = leaves.len();
+    let proof_flags_len = proof_flags.len();
+
+    // Check proof validity.
+    if (leaves_len + proof.len() != proof_flags_len + 1) {
+        panic!("MerkleProof: invalid multi proof");
+    }
+
+    // The x_pos values are "pointers" to the next value to consume in each array.
+    // By incrementing the value we simulate a queue's pop operation.
+    let mut hashes = array![];
+    let mut leaf_pos = 0;
+    let mut hash_pos = 0;
+    let mut proof_pos = 0;
+    let mut i = 0;
+
+    // At each step, we compute the next hash using two values:
+    // - a value from the "main queue". If not all leaves have been consumed, we get the next leaf,
+    // otherwise we get the next hash.
+    // - depending on the flag, either another value from the "main queue" (merging branches) or an
+    // element from the `proof` array.
+    while i < proof_flags_len {
+        let a = if leaf_pos < leaves_len {
+            leaf_pos += 1;
+            leaves.at(leaf_pos - 1)
+        } else {
+            hash_pos += 1;
+            hashes.at(hash_pos - 1)
+        };
+
+        let b = if *proof_flags.at(i) {
+            if leaf_pos < leaves_len {
+                leaf_pos += 1;
+                leaves.at(leaf_pos - 1)
+            } else {
+                hash_pos += 1;
+                hashes.at(hash_pos - 1)
+            }
+        } else {
+            proof_pos += 1;
+            proof.at(proof_pos - 1)
+        };
+
+        hashes.append(Hasher::commutative_hash(*a, *b));
+        i += 1;
+    };
+
+    let root = if proof_flags_len > 0 {
+        // If `proof_flags` is not empty, assert that every
+        // proof was used in the validation process.
+        if proof_pos != proof.len() {
+            // TODO: check if this is not unrechable code.
+            panic!("MerkleProof: invalid multi proof");
```

Comment:
Will `assert_eq` fit nicer here?

Reply:
Sadly `assert_eq` only works in tests, and using assert restricts the message to a shortstring.